server/src/code-runner/docker-runner.ts

```diff
@@ -91,16 +123,9 @@ export class DockerRunner implements CodeRunner {
               })
             ),
             concat(this.config.downloader.fetch(containerId, configuration.inputs)),
-            concat(
-              this.config.spawn(this.config.dockerExecutable, [
-                'exec',
-                '-t',
-                containerId,
-                '/bin/bash',
-                '-c',
-                `mkdir /run/outputs && cd /run && python main.py ${args}`
-              ])
-            ),
+            // Only install `requirements.txt` when the container is writeable
+            concat(this.runCommand(containerId, 'cd /run && test -f requirements.txt && pip install -r requirements.txt', environment.writeable)),
+            concat(this.runCommand(containerId, `mkdir /run/outputs && cd /run && python main.py ${args}`)),
```

Comment:
Just throwing this out here (now that I see why the conditional `runCommand()`:

```
switchMap(_ => environment.writable ? /* runInstallCommand */ : of(null)),
contact(/* runRunCommand */)
```

Reply:
Is that correct? Because we need to `concat` all the output.